plumb.1
.Dd July 2, 2023
.Dt PLUMB 1
.Os
.Sh NAME
.Nm plumb
.Nd run best command for given arguments
.Sh SYNOPSIS
.Nm
.Op Fl Ar action ...
.Op Cm "--"
.Ar arg ...
.Sh DESCRIPTION
.Nm
passes the given arguments, as is or modified,
to the best command associated to the rules the arguments match with.
.Pp
.Nm
can be used, for example, to open files or URLs.
By giving filenames as arguments,
.Nm
tries to find the best command to open them
after matching the filenames with sets of rules in sequence.
See the section
.Sx "EXAMPLES"
for illustration.
.Pp
Each set of rule (or ruleset) is associated with a command for a given type of action
(such as
.Qq "open"
or
.Qq "edit" ) .
The command for the given action associated with the first ruleset
matching the arguments is the one chosen to be run.
.Pp
The first arguments beginning with hyphen
.Pq Cm "-" ,
are interpreted as a actions to be try.
If the matching ruleset contains one of those actions,
the first action is performed.
For example,
.Fl edit Fl open
will try to perform either the edit or the open action on the following arguments.
.Pp
The actions
.Fl o
and
.Fl e
are equivalent to
.Fl open
and
.Fl edit ,
respectively.
.Pp
A
.Cm "--"
separates actions from actual arguments.
.Pp
If no action is provided,
.Nm
acts as if the action
.Cm "-open"
were given.
.Sh USAGE
.Nm
reads rules from the file
.Pa "$HOME/lib/plumb" .
Lines with blank and lines beginning with
.Qq "#" are ignored.
.Pp
Each non ignored line can be either a static variable assignment
or can be one of the five following types,
named after the second word in the line:
.Bl -tag -width Ds
.It Qq "for"
A line beginning a ruleset.
.It Qq "matches"
A line describing a condition a variable has to match case-sensitively,
and, optionally, setting new variables when the matching occurs.
.It Qq "imatches"
A line describing a condition a variable has to match case-insensitively,
and, optionally, setting new variables when the matching occurs.
.It Qq "types"
A line testing the existence and type of a file,
and assigning this information to a new variable.
.It Qq "at"
A line testing the existence of a file,
and assigning its absolute path to a new variable.
.It Qq "with"
A line describing the command to be open if the conditions in the same ruleset matches.
.El
.Pp
There are two kinds of variables that can be assigned and used in the configuration file:
.Bl -tag -width Ds
.It Static variables
Static variables are assigned with a
.Qq "NAME=VALUE"
line.
Such variables are expanded into a single word
when prefixed with a dollar sign
.Pq "$"
and occurring in any line after the place it was defined.
Environment variables are of this kind, but do not need to be defined.
See the section
.Sx "Static variables"
for more information on static variables.
.It Argument variables
Argument variables are assigned with a
.Qq "matches" ,
.Qq "imatches" ,
.Qq "types" ,
or a
.Qq "at"
line.
Such variables can be expanded into various words
(one for each argument passed to
.Nm )
when prefixed with a percent sign
.Pq "%"
and occurring as the last argument of a
.Qq "with"
line in the same ruleset it was defined.
See the section
.Sx "WITH-lines"
for more information on argument variables.
.El
.Pp
Each line is a sequence of words
(which are either space-delimited words or
strings quoted in
.Xr rc 1
single-quote style).
The first word of a line is the
.Qq "subject" .
The second word (which identifies the type of the line) is the
.Qq "verb" .
The remaining words are the arguments.
.Pp
The configuration is processed once for each argument.
For each processing pass, the argument variable
.Ic data
is set to the argument itself
and the remaining argument variables are re-assigned.
Static variables are assigned only once, at
.Nm
initialization.
.Ss Static variables
Lines of the form
.Qq "NAME=VALUE"
assign a value to a static variable.
Static variable are recognized anywhere in the file after the place they are defined.
.Pp
Environment variables are also static variables, but they are not defined in the config file
(they are already defined in the environment).
.Pp
References to static variables can occur on the configuration file outside quotations,
and are replaced with their values.
Such references are prefixed with the dollar sign
.Qq "$"
or prefixed with the dollar sign and surrounded by curly braces.
The dollar sign can be escaped by doubling it.
.Ss FOR-lines
Lines whose second word is
.Qq "for"
must have
.Qq "rules"
as subject.
They begin a new ruleset.
A ruleset is everything between a
.Qq FOR-line
and the next one.
.Pp
For example, the following line begins the ruleset for handling video files:
.Bd -literal -offset indent
rules for video files
.Ed
.Pp
The arguments of a
.Qq FOR-line
are the name of the ruleset.
.Pp
Conditions in a ruleset are only checked within the ruleset.
Variables set in a ruleset are only valid within the ruleset.
.Pp
The lines before the first
.Qq FOR-line
make the global, unamed ruleset.
Conditions in the global ruleset are ignored.
Variables set in the global ruleset are valid for the entire file.
.Ss MATCHES-lines
Lines whose second word is
.Qq "matches"
or
.Qq "imatches"
must have the name of a variable as subject (first word in the line).
They must also have a regular expression as first argument.
The subject names a value that must match the regular expression.
.Pp
If the second word is
.Qq "matches" ,
the regular expression matching is case-sensitive.
If the second word is
.Qq "imatches" ,
the regular expression matching is case-insensitive.
.Pp
For example, the following line is a three-word condition that
says that one of the conditions for the current ruleset to be matched
is for the content of the variable
.Ic mime
to match the regular expression
.Ic "image/(jpeg|png)" .
.Bd -literal -offset indent
mime matches 'image/(jpeg|png)'
.Ed
.Pp
The regular expression is a extended POSIX regular expression
and must match the entire value of the argument variable for the condition to be valid.
.Pp
If the rule has more than one argument, the second argument must be
.Ic into
and the following ones must be names of argument variables to be set.
Each argument variable is set to the substring matching the parenthesized subexpression
of the regular expression if, and only if, the full regular expression matches
the value of the subject.
.Pp
For example, the following line assigns to the argument variable
.Ic base
the
.Xr basename 1
of the value on the argument variable
.Ic data
(supposing it contains a filename);
and assigns to the argument variable
.Ic extension
the extension of the filename.
If either subexpression does not match, the corresponding argument variable is
set to the empty string.
The dummy argument variable name
.Ic _
(underscore) is used for uneeded values.
.Bd -literal -offset indent
data matches '(([^/]*/)*)([^/]*(\e.([A-Za-z0-9]+)?))' _ _ base extension
.Ed
.Ss TYPES-lines
Lines whose second word is
.Qq "types"
must have the name of an argument variable as subject,
and the name of another argument variable as its sole argument.
The subject names a value for an existing file whose mimetype is assigned
to the argument variable passed as argument.
.Pp
For example, the following line is a three-word assignment that says
that the mimetype of the file named in the argument variable
.Ic "data"
must be assigned to the argument variable
.Ic "mime" .
.Bd -literal -offset indent
data types mime
.Ed
.Ss AT-lines
Lines whose second word is
.Qq "at"
must have the name of an argument variable as subject,
and the name of another argument variable as its sole argument.
The subject names a value for an existing file whose real absolute path is assigned
to the argument variable passed as argument.
.Pp
For example, the following line is a three-word assignment that says
that the absolute path of the file named in the argument variable
.Ic "data"
must be assigned to the argument variable
.Ic "path" .
.Bd -literal -offset indent
data at path
.Ed
.Ss WITH-lines
Lines whose second word is
.Qq "with"
must have the name of an action type
(like
.Ic "open"
or
.Ic "edit" )
as subject and a command invocation as arguments.
The arguments name a program to be run for the action named as subject
when the ruleset the line is in is valid for all the arguments passed.
.Pp
for example, the following line is a three-word description to open the browser
.Xr firefox 1
on the
.Ic open
action.
.Bd -literal -offset indent
open with firefox
.Ed
.Pp
If the last argument has a percent symbol
.Pq Qq "%"
before a name,
then this name is considered as a variable name.
This argument is replaced by one argument for each argument passed
and the variable name with the percent sign is replaced with the value of the variable.
.Pp
For example, the following line opens
.Xr firefox 1
replacing the argument
.Ic "file://%data"
for the variable
.Ic "data"
for each argument.
(so if
.Nm
is invoked for
.Pa "./index.html"
and
.Pa "/path/to/file.html" ,
then that single argument is replaced with
.Pa "file://./index.html"
and
.Pa "file:///path/to/file.html" ) .
.Bd -literal -offset indent
open with firefox -- file://%data
.Ed
.Pp
Just like environment variables, the percent sign can be escaped by doubling it.
The name of the variable can also occur between curly braces.
.Sh ENVIRONMENT
The following environment variables affect the execution of
.Nm .
.Bl -tag -width Ds
.It Ev HOME
Path to the directory to search for the file
.Pa "lib/plumb" .
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa "$HOME/lib/plumb"
.Nm Ns 's
default configuration file.
.El
.Sh EXIT STATUS
.Ex -std
.Pp
It is an error if no ruleset matches for an argument.
.Sh EXAMPLES
The following is the example of a simple configuration file.
.Bd -literal -offset indent
DATAREGEX = '(([A-Za-z]+):(//)?)?(.*(\e.([A-Za-z0-9]+))?)'

data      matches     $DATAREGEX into _ protocol _ file _ extension
file      types       mime
file      at          path

rules     for         youtube video
protocol  matches     '(ytdl|https?)?'
file      matches     '(.*/)?[A-Za-z0-9_-]{11}'
open      with        mpv --force-window=immediate -- %data

rules     for         image file
protocol  matches     '(file)?'
mime      imatches    'image/(png|jpe?g|tiff)'
open      with        display -- %path
edit      with        gimp -- %file

rules     for         web page
protocol  matches     '(https?|file)?'
extension imatches    'html'
open      with        seamonkey -- %data
.Ed
.Pp
This configuration file is interpreted as follows:
.Bl -bullet
.It
The static variable
.Ic DATAREGEX
is set to a regular expression used later in the config file.
.It
For each passed argument, the second paragraph sets the argument variables
.Qq Ic protocol
to an URI protocol;
.Qq Ic file
to the argument without the protocol;
.Qq Ic extension
to a file extension; and
.Qq Ic mime
to the mimetype of the value of
.Qq Ic file .
The argument variable
.Qq Ic data
is automatically set to the argument itself on each pass.
.It
The third paragraph sets rules for opening youtube videos on
.Xr mpv 1
using the
.Ic ytdl
protocol.
.It
The fourth paragraph sets rules for opening and editing image files.
.It
The fifth paragraph sets rules for opening web pages.
.El
.Pp
With this configuration file, the following command opens
.Em https://wikipedia.org
and
.Em file:///var/www/htdocs/index.html
on
.Xr seamonkey 1 :
.Bd -literal -offset indent
$ plumb https://wikipedia.org file:///var/www/htdocs/index.html
.Ed
.Pp
The following command opens a PNG file on gimp for editing:
.Bd -literal -offset indent
$ plumb -edit /home/user/photo.png
.Ed
.Sh SEE ALSO
.Rs
.%A "Rob Pike"
.%T "Plumbing and Other Utilities"
.%I "Bell Laboratories"
.Re
.Sh HISTORY
A
.Nm
utility appeared in the Plan 9 operating system.


plumb.c
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <assert.h>
#include <err.h>
#include <limits.h>
#include <regex.h>
#include <spawn.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

#include <magic.h>

#define LEN(a)          (sizeof(a) / sizeof((a)[0]))
#define ENVVAR_MAX      128
#define MAXTOKENS       128
#define RULESPATH       "/lib/plumb"
#define HOME            "HOME"
#define OPEN_ACTION     "open"
#define EDIT_ACTION     "edit"
#define DEF_ACTION      OPEN_ACTION
#define ALPHANUM        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_"

struct Variable {
	struct Variable *next;
	char *name;
	char *value;
} *globals, *locals;

struct Argument {
	struct Variable *globals, *locals;
	char *data;
};

struct Ruleset {
	/*
	 * Each block of rule, initiated by a "rules for <RULESET_NAME>"
	 * line on the config file, is a ruleset.  A ruleset is made of
	 * its name (as an array of arguments) and a list of rules.
	 *
	 * Each rule represents a line in the config file with the form
	 * "<SUBJ> <TYPE> <ARGS>".  <SUBJ> is a string that dictates the
	 * rule's subject; <TYPE> is the string "matches", "types" or
	 * "with"; and <ARGS> is an array of argument strings.
	 *
	 * A rule of type "matches" has a special argument.  Its first
	 * argument is a regular expression, while the following
	 * optional arguments are strings.
	 */
	struct Ruleset *next;
	struct Rule {
		struct Ruleset *set;
		struct Rule *next;
		char *subj;
		enum Type {
			RULE_MATCHES,
			RULE_TYPES,
			RULE_WITH,
			RULE_AT,
		} type;
		regex_t reg;            /* only used in "matches" rules */
		char **argv;
		size_t argc;
	} *rules;
	char **argv;
	size_t argc;
};

struct Parsectx {
	/*
	 * Context for parsing the config file.
	 */
	FILE *fp;
	char *filename;
	size_t lineno;
	bool goterror;
};

extern char **environ;

static magic_t magic;

static void
usage(void)
{
	(void)fprintf(stderr, "usage: plumb [-actions] arg...\n");
	exit(EXIT_FAILURE);
}

static void *
erealloc(void *p, size_t size)
{
	if ((p = realloc(p, size)) == NULL)
		err(EXIT_FAILURE, "realloc");
	return p;
}

static void *
ecalloc(size_t nmemb, size_t size)
{
	void *p;

	if ((p = calloc(nmemb, size)) == NULL)
		err(EXIT_FAILURE, "calloc");
	return p;
}

static void *
emalloc(size_t size)
{
	void *p;

	if ((p = malloc(size)) == NULL)
		err(EXIT_FAILURE, "malloc");
	return p;
}

static char *
estrdup(const char *s)
{
	char *p;

	if ((p = strdup(s)) == NULL)
		err(EXIT_FAILURE, "strdup");
	return p;
}

static char *
estrndup(const char *s, size_t maxlen)
{
	char *p;

	if ((p = strndup(s, maxlen)) == NULL)
		err(EXIT_FAILURE, "strndup");
	return p;
}

static pid_t
efork(void)
{
	pid_t pid;

	if ((pid = fork()) < 0)
		err(EXIT_FAILURE, "fork");
	return pid;
}

static int
beginsruleset(char **toks, size_t ntoks)
{
	return ntoks > 2 &&
		strcmp(toks[0], "rules") == 0 &&
		strcmp(toks[1], "for") == 0;
}

static void
syntaxerr(struct Parsectx *parse, const char *msg, const char *arg)
{
	if (arg == NULL) {
		warnx(
			"%s:%zu: syntax error: %s",
			parse->filename,
			parse->lineno,
			msg
		);
	} else {
		warnx(
			"%s:%zu: syntax error: %s \"%s\"",
			parse->filename,
			parse->lineno,
			msg,
			arg
		);
	}
	parse->goterror = true;
}

static struct Rule *
newrule(struct Parsectx *parse, char *toks[], size_t ntoks)
{
	struct Rule *rule;
	regex_t reg;
	size_t i, n;
	int flags, res;
	enum Type type;
	char errbuf[1024];

	assert(ntoks > 2);      /* we have handled ntoks <= 2 before */
	flags = REG_EXTENDED;
	if (strcmp(toks[1], "matches") == 0) {
		type = RULE_MATCHES;
	} else if (strcmp(toks[1], "imatches") == 0) {
		flags |= REG_ICASE;
		type = RULE_MATCHES;
	} else if (strcmp(toks[1], "types") == 0) {
		type = RULE_TYPES;
	} else if (strcmp(toks[1], "with") == 0) {
		type = RULE_WITH;
	} else if (strcmp(toks[1], "at") == 0) {
		type = RULE_AT;
	} else {
		syntaxerr(parse, "unknown predicate", toks[1]);
		return NULL;
	}
	memset(&reg, 0, sizeof(reg));
	n = 2;
	switch (type) {
	case RULE_MATCHES:
		n++;
		if (ntoks > 3) {
			if (strcmp(toks[3], "into") != 0) {
				syntaxerr(parse, "unknown argument", toks[3]);
				return NULL;
			}
			n++;
		}
		if ((res = regcomp(&reg, toks[2], flags)) != 0) {
			if (regerror(res, &reg, errbuf, sizeof(errbuf)) > 0)
				syntaxerr(parse, errbuf, NULL);
			else
				syntaxerr(parse, "wrong regex", toks[2]);
			regfree(&reg);
			return NULL;
		}
		break;
	case RULE_TYPES:
	case RULE_AT:
		if (ntoks != 3) {
			syntaxerr(parse, "improper number of arguments", NULL);
			return NULL;
		}
		break;
	case RULE_WITH:
		/* we deal with this later */
		break;
	}
	rule = emalloc(sizeof(*rule));
	*rule = (struct Rule){
		.set = NULL,
		.type = type,
		.reg = reg,
		.argv = NULL,
		.argc = 0,
		.subj = toks[0],
	};
	if (n < ntoks) {
		rule->argc = ntoks - n;
		rule->argv = ecalloc(rule->argc, sizeof(*rule->argv));
		memcpy(rule->argv, toks + n, rule->argc * sizeof(*rule->argv));
	}
	for (i = 1; i < n; i++)
		free(toks[i]);
	return rule;
}

static void
insertvar(struct Variable **head, char *name, char *value)
{
	struct Variable *var;

	if (name == NULL) {
		free(value);
		return;
	}
	for (var = *head; var != NULL; var = var->next) {
		if (strcmp(var->name, name) == 0) {
			free(var->value);
			var->value = value;
			return;
		}
	}
	var = emalloc(sizeof(*var));
	*var = (struct Variable){
		.name = name,
		.value = value,
	};
	var->next = *head;
	*head = var;
}

static char *
lookupvar(struct Variable *var, const char *name)
{
	if (name == NULL)
		return NULL;
	for ( ; var != NULL; var = var->next)
		if (strcmp(var->name, name) == 0)
			return var->value != NULL ? var->value : "";
	return getenv(name);
}

static size_t
strnrspn(char *buf, char *charset, size_t len)
{
	while (len > 0 && strchr(charset, buf[len - 1]) != NULL)
		len--;
	return len;
}

static size_t
tokenize(struct Parsectx *parse, struct Variable *env,
         char *str, char *toks[], size_t maxtoks)
{
	char *buf, *val;
	size_t ntoks, bufsize, len;
	size_t j, k;
	char c;
	bool inquote = false;

	ntoks = 0;
	bufsize = 0;
	buf = NULL;
	while (*str != '\0' && ntoks < maxtoks) {
		str += strspn(str, " \t");
		j = 0;
		while (*str != '\0') {
			if (!inquote && strchr(" \t", *str))
				break;
			val = NULL;
			if (!inquote && str[0] == '$' && str[1] == '{') {
				/* ${VARIABLE} */
				str += 2;
				k = strcspn(str, "}");
				if (k == 0) {
					syntaxerr(parse, "bad substitution", "${}");
					goto error;
				}
				if (str[k] != '}') {
					syntaxerr(parse, "unmatching bracket", NULL);
					goto error;
				}
				str[k] = '\0';
				if ((val = lookupvar(env, str)) == NULL)
					val = "";
				len = strlen(val);
				str += k + 1;
			} else if (!inquote && str[0] == '$' && str[1] == '$') {
				/* $$ */
				str += 2;
				val = "$";
				len = 1;
			} else if (!inquote && str[0] == '$') {
				/* $VARIABLE */
				str++;
				k = strspn(str, ALPHANUM);
				if (k == 0) {
					syntaxerr(parse, "bad substitution", "$");
					goto error;
				}
				c = str[k];
				str[k] = '\0';
				if ((val = lookupvar(env, str)) == NULL)
					val = "";
				len = strlen(val);
				str += k;
				if (c != '\0') {
					str[0] = c;
				}
			} else if (inquote && str[0] == '\'' && str[1] == '\'') {
				/* '' */
				str += 2;
				val = "'";
				len = 1;
			} else if (str[0] == '\'') {
				/* ' */
				inquote = !inquote;
				str++;
				continue;
			} else {
				/* char */
				val = str;
				len = 1;
				str++;
			}
			if (j + len + 1 > bufsize) {
				bufsize += len + 128;
				buf = erealloc(buf, bufsize);
			}
			memcpy(buf + j, val, len);
			j += len;
		}
		if (buf == NULL)
			return 0;
		buf[j] = '\0';
		toks[ntoks++] = estrdup(buf);
	}
	free(buf);
	return ntoks;
error:
	free(buf);
	for (j = 0; j < ntoks; j++)
		free(toks[j]);
	return 0;
}

static void
freevars(struct Variable *head, bool freename)
{
	struct Variable *tmp;

	while (head != NULL) {
		tmp = head;
		head = head->next;
		if (freename)
			free(tmp->name);
		free(tmp->value);
		free(tmp);
	}
}

static struct Ruleset *
readrules(struct Parsectx *parse)
{
	struct Rule *rule = NULL, *currule;
	struct Ruleset *head = NULL, *curset, *set;
	struct Variable *env = NULL;
	ssize_t linelen;
	size_t ntoks, n, i;
	size_t linesize = 0;
	char *toks[MAXTOKENS];
	char *line = NULL;
	char *str = NULL;

	head = curset = emalloc(sizeof(*curset));
	*curset = (struct Ruleset){ 0 };
	while ((linelen = getline(&line, &linesize, parse->fp)) != -1) {
		parse->lineno++;
		str = line;
		n = strspn(line, " \t");
		str += n;
		linelen -= n;
		str[strnrspn(line, " \t\n", linelen)] = '\0';
		if (str[0] == '\0' || str[0] == '#')
			continue;
		if ((ntoks = tokenize(parse, env, str, toks, LEN(toks))) == 0)
			continue;
		if (ntoks < 3) {
			syntaxerr(parse, "improper rule", NULL);
			goto error;
		}
		if (toks[1][0] == '=' && toks[1][1] == '\0') {
			/* variable assignment */
			if (ntoks != 3) {
				syntaxerr(parse, "bad assignment", NULL);
				goto error;
			}
			insertvar(&env, toks[0], toks[2]);
			free(toks[1]);         /* free "=" */
			continue;
		} else if (beginsruleset(toks, ntoks)) {
			/* a new ruleset begins */
			set = emalloc(sizeof(*set));
			*set = (struct Ruleset){
				.next = NULL,
				.rules = NULL,
				.argv = NULL,
				.argc = ntoks - 2,
			};
			set->argv = ecalloc(set->argc, sizeof(*set->argv));
			memcpy(set->argv, toks+2, set->argc*sizeof(*set->argv));
			free(toks[0]);          /* free "rules" */
			free(toks[1]);          /* free "for" */
			curset->next = set;
			curset = set;
		} else if ((rule = newrule(parse, toks, ntoks)) != NULL) {
			/* new rule for current ruleset */
			rule->set = curset;
			if (curset->rules == NULL)
				curset->rules = rule;
			else
				currule->next = rule;
			currule = rule;
		} else {
			syntaxerr(parse, "improper rule", NULL);
error:
			for (i = 0; i < ntoks; i++) {
				free(toks[i]);
			}
		}
	}
	freevars(env, true);
	free(line);
	return head;
}

static char *
getconfig(void)
{
	char *home, *filename;
	size_t size;

	if ((home = getenv(HOME)) == NULL)
		errx(EXIT_FAILURE, "could not find $HOME");
	size = strlen(home) + sizeof(RULESPATH);
	filename = emalloc(size);
	(void)snprintf(filename, size, "%s" RULESPATH, home);
	return filename;

}

static void
freerules(struct Ruleset *head)
{
	struct Ruleset *set;
	struct Rule *rule;
	size_t i;

	while (head != NULL) {
		set = head;
		head = set->next;
		while (set->rules != NULL) {
			rule = set->rules;
			set->rules = rule->next;
			if (rule->type == RULE_MATCHES)
				regfree(&rule->reg);
			for (i = 0; i < rule->argc; i++)
				free(rule->argv[i]);
			free(rule->argv);
			free(rule->subj);
			free(rule);
		}
		for (i = 0; i < set->argc; i++)
			free(set->argv[i]);
		free(set->argv);
		free(set);
	}
}

static char *
lookupargvar(struct Variable *globals, struct Variable *locals,
             char *data, char *name)
{
	char *value;

	value = lookupvar(locals, name);
	if (value == NULL)
		value = lookupvar(globals, name);
	if (value == NULL && strcmp("data", name) == 0)
		value = data;
	if (value == NULL)
		value = "";
	return value;
}

static struct Rule *
matchruleset(struct Ruleset *set, char *arg,
             char **actions, int nactions,
             struct Variable *globals, struct Variable **locals_ret)
{
	struct Variable *locals = NULL;
	struct Rule *rule = NULL;
	struct stat sb;
	regmatch_t pmatch[MAXTOKENS];
	regoff_t beg, len;
	size_t i;
	int j;
	char *value, *newstr;
	const char *filetype;
	bool match = true;

	for (rule = set->rules; rule != NULL; rule = rule->next) {
		if (rule->type == RULE_WITH)
			continue;       /* we handle RULE_WITH later */
		value = lookupargvar(globals, locals, arg, rule->subj);
		if (rule->type == RULE_TYPES) {
			if (stat(value, &sb) == -1)
				filetype = NULL;
			else if (S_ISDIR(sb.st_mode))
				filetype = "inode/directory";
			else
				filetype = magic_file(magic, value);
			if (filetype != NULL) {
				newstr = estrdup(filetype);
				insertvar(&locals, rule->argv[0], newstr);
			} else {
				insertvar(&locals, rule->argv[0], NULL);
			}
			continue;
		}
		if (rule->type == RULE_AT) {
			insertvar(
				&locals,
				rule->argv[0],
				realpath(value, NULL)   /* freed later */
			);
			continue;
		}
		/* rule->type == RULE_MATCHES */
		if (regexec(&rule->reg, value, MAXTOKENS, pmatch, 0) != 0) {
			match = false;
			continue;
		}
		if (pmatch[0].rm_so != 0 || value[pmatch[0].rm_eo] != '\0') {
			match = false;
			continue;
		}
		for (i = 0; i < rule->reg.re_nsub && i < rule->argc; i++) {
			beg = pmatch[i + 1].rm_so;
			len = pmatch[i + 1].rm_eo - beg;
			if (beg >= 0 && len >= 0) {
				newstr = estrndup(value + beg, len);
				insertvar(&locals, rule->argv[i], newstr);
			} else {
				insertvar(&locals, rule->argv[i], NULL);
			}
		}
	}
	if (locals_ret != NULL)
		*locals_ret = locals;
	else
		freevars(locals, false);
	if (!match)
		return NULL;
	for (rule = set->rules; rule != NULL; rule = rule->next) {
		if (rule->type != RULE_WITH)
			continue;
		for (j = 0; j < nactions; j++) {
			if (strcmp(actions[j], rule->subj) == 0) {
				return rule;
			}
		}
	}
	return NULL;
}

static void
plumb(struct Rule *rule, struct Argument *args, int argc)
{
	char **newargv = NULL;
	char **cmd, **p;
	char *buf = NULL;
	char *str, *var, *val;
	size_t len, k;
	size_t pos = 0;
	size_t size;
	size_t bufsize = 0;
	int newargc = 0;
	int i;
	char ch;
	bool gotvar = false;

	if (rule == NULL) {
		warnx("could not find rule for arguments");
		return;
	}

	/*
	 * Anounce the ruleset we are plumbing.
	 */
	(void)fprintf(stderr, "plumbing ");
	for (k = 0; k < rule->set->argc; k++) {
		(void)fprintf(
			stderr,
			"%s%s",
			(k == 0 ? "" : " "),
			rule->set->argv[k]
		);
	}
	(void)fprintf(stderr, "\n");

	/*
	 * Find the substring "%var" in argv[i], where i is the last
	 * element of argv[].
	 */
	str = rule->argv[rule->argc - 1];
	len = 0;
	var = NULL;
	while (*str != '\0') {
		if (!gotvar && str[0] == '%' && str[1] == '{') {
			/* %{VARIABLE} */
			k = strcspn(str + 2, "}");
			if (k == 0 || str[k + 2] != '}')
				goto fallback;
			str += 2;
			str[k] = '\0';
			var = str;
			str += k + 1;
			gotvar = true;
			pos = len;
			continue;
		} else if (!gotvar && str[0] == '%') {
			/* %VARIABLE */
			k = strspn(str + 1, ALPHANUM);
			if (k == 0)
				goto fallback;
			str++;
			ch = str[k];
			str[k] = '\0';
			var = str;
			str += k;
			if (ch != '\0')
				str[0] = ch;
			pos = len;
			gotvar = true;
			continue;
		} else if (!gotvar && str[0] == '%' && str[1] == '%') {
			/* %% */
			ch = '%';
			str += 2;
		} else {
fallback:
			/* char */
			ch = *str;
			str++;
		}
		if (len + 2 > bufsize) {
			bufsize += len + 128;
			buf = erealloc(buf, bufsize);
		}
		buf[len++] = ch;
	}
	if (buf != NULL) {
		buf[len] = '\0';
		str = buf;
	} else {
		str = "";
		pos = len = 0;
	}

	/*
	 * Create the command cmd[] to be spawned.
	 *
	 * If rule->argv[rule->argc - 1] (the last argument of the
	 * plumbing rule) does not contain the "%var" substring, cmd[]
	 * is exactly the plumbing rule itself (cmd == rule->argv).
	 *
	 * However, if rule->argv[rule->argc - 1] (the last argument of
	 * the plumbing rule) contains the "%var" substring, cmd[] is
	 * composed of:
	 * - The strings rule->argv[0] to rule->argv[rule->argc - 2].
	 * - n strings (n is the number of command-line arguments given
	 *   to plumb), each one equal to rule->argv[rule->argc - 1],
	 *   but with the "%var" substring replaced with the value that
	 *   "%var" has for each command-line argument.
	 *
	 * For example, if plumb is invoked as
	 *
	 *      $ plumb -edit foo.png file:///var/www/bar.jpg
	 *
	 * And the rule for editing image files is
	 *
	 *      gimp -s -- %path
	 *
	 * And the value of the variable %path for the first argument is
	 *
	 *      foo.png
	 *
	 * And the value of %path for the second argument is
	 *
	 *      /var/www/bar.jpg
	 *
	 * Then, the command to be executed is
	 *
	 *      gimp -s -- foo.png /var/www/bar.jpg
	 */
	p = NULL;
	cmd = rule->argv;
	if (var != NULL) {
		newargc = argc,
		newargv = ecalloc(newargc, sizeof(*newargv));
		for (i = 0; i < argc; i++) {
			val = lookupargvar(
				args[i].globals,
				args[i].locals,
				args[i].data,
				var
			);
			size = len + strlen(val) + 1;
			newargv[i] = emalloc(size);
			(void)snprintf(
				newargv[i],
				size,
				"%.*s%s%.*s",
				(int)pos,
				str,
				val,
				(int)(len - pos),
				str + pos
			);
		}
		p = ecalloc(rule->argc + newargc, sizeof(*p));
		cmd = p;
		for (k = 0; k < rule->argc - 1; k++)
			cmd[k] = rule->argv[k];
		for (k = 0; k < (size_t)newargc; k++)
			cmd[rule->argc - 1 + k] = newargv[k];
		cmd[rule->argc + newargc - 1] = NULL;
	}

	if (efork() == 0) {
		if (posix_spawnp(NULL, cmd[0], NULL, NULL, cmd, environ))
			err(EXIT_FAILURE, "posix_spawnp");
		exit(EXIT_SUCCESS);
	}
	free(p);
	free(buf);
	for (i = 0; i < newargc; i++)
		free(newargv[i]);
	free(newargv);
}

static void
freeargs(struct Argument *args, int argc)
{
	int i;

	for (i = 0; i < argc; i++) {
		freevars(args[i].globals, false);
		freevars(args[i].locals, false);
	}
	free(args);
}

int
main(int argc, char *argv[])
{
	struct Ruleset *sets, *set;
	struct Rule *plumbwith = NULL;
	struct Rule *newaction = NULL;
	struct Argument *args;
	struct Parsectx parse = { 0 };
	size_t span;
	char **actions;
	int nactions;
	int i;

	actions = &argv[1];
	nactions = 0;
	for (i = 1; i < argc; i++) {
		span = strspn(argv[i], "-");
		if (span == 0)                       /* argv[i] is not -word */
			break;
		if (argv[i][span] == '\0') {         /* argv[i] is -- */
			i++;
			break;
		}
		if (strcmp(argv[i], "-o") == 0)      /* argv[i] is -o */
			argv[i] = OPEN_ACTION;
		else if (strcmp(argv[i], "-e") == 0) /* argv[i] is -e */
			argv[i] = EDIT_ACTION;
		else                                 /* argv[i] is -word */
			argv[i]++;
		nactions++;
	}
	argc -= i;
	argv += i;
	if (argc == 0)
		usage();
	if (nactions == 0) {
		actions = (char *[]){ DEF_ACTION };
		nactions = 1;
	}
	args = ecalloc(argc, sizeof(*args));
	magic = magic_open(
		MAGIC_SYMLINK | MAGIC_MIME_TYPE |
		MAGIC_PRESERVE_ATIME | MAGIC_ERROR
	);
	if (magic == NULL)
		errx(EXIT_FAILURE, "could not get magic cookie");
	if (magic_load(magic, NULL) == -1)
		errx(EXIT_FAILURE, "could not load magic database");
	parse.filename = getconfig();
	if ((parse.fp = fopen(parse.filename, "r")) == NULL)
		err(EXIT_FAILURE, "%s", parse.filename);
	sets = readrules(&parse);
	for (i = 0; i < argc; i++) {
		/*
		 * First we run on the top ruleset, which should contain
		 * no action rule, so we can fill in the global argument
		 * variables.
		 */
		args[i].data = argv[i];
		args[i].globals = NULL;
		args[i].locals = NULL;
		set = sets;
		(void)matchruleset(
			set, argv[i],
			actions, nactions,
			NULL, &args[i].globals
		);
		if (i == 0) {
			/*
			 * For the first argument, we find a ruleset
			 * matching it.
			 */
			while ((set = set->next) != NULL) {
				plumbwith = matchruleset(
					set, argv[i],
					actions, nactions,
					args[i].globals, &args[i].locals
				);
				if (plumbwith != NULL)
					break;
				freevars(args[i].locals, false);
				args[i].locals = NULL;
			}
			if (plumbwith == NULL) {
				break;
			}
		} else {
			/*
			 * For the following arguments, we check whether
			 * the it matchs the ruleset of the 1st argument.
			 */
			newaction = matchruleset(
				plumbwith->set, argv[i],
				actions, nactions,
				args[i].globals, &args[i].locals
			);
			if (plumbwith != newaction) {
				freevars(args[i].locals, false);
				args[i].locals = NULL;
				plumbwith = NULL;
				break;
			}
		}
	}
	magic_close(magic);
	plumb(plumbwith, args, argc);
	freerules(sets);
	free(parse.filename);
	freeargs(args, argc);
	return 0;
}
